core/lib/object_store/src/s3.rs

```diff
@@ -0,0 +1,225 @@
+                is_retriable: false,
+            },
+            SdkError::DispatchFailure(_) | SdkError::TimeoutError(_) => ObjectStoreError::Other {
+                source: err.into(),
+                is_retriable: true,
+            },
+            SdkError::ResponseError(http_response) => {
+                let code = http_response.raw().status().as_u16();
+                if code == StatusCode::NOT_FOUND.as_u16() {
+                    ObjectStoreError::KeyNotFound(err.into())
+                } else if code == StatusCode::FORBIDDEN.as_u16() {
+                    ObjectStoreError::Initialization {
+                        source: err.into(),
+                        is_retriable: false,
+                    }
+                } else {
+                    ObjectStoreError::Other {
+                        source: err.into(),
+                        is_retriable: true,
+                    }
+                }
+            }
+            SdkError::ServiceError(http_response) => {
+                let code = http_response.raw().status().as_u16();
+                if code == StatusCode::NOT_FOUND.as_u16() {
+                    ObjectStoreError::KeyNotFound(err.into())
+                } else if code == StatusCode::FORBIDDEN.as_u16() {
+                    ObjectStoreError::Initialization {
+                        source: err.into(),
+                        is_retriable: false,
+                    }
+                } else {
+                    ObjectStoreError::Other {
+                        source: err.into(),
+                        is_retriable: true,
+                    }
+                }
+            }
+            _ => ObjectStoreError::Other {
+                source: err.into(),
+                is_retriable: false,
+            },
+        }
+    }
+}
+
+#[async_trait]
+impl ObjectStore for S3Store {
+    async fn get_raw(&self, bucket: Bucket, key: &str) -> Result<Vec<u8>, ObjectStoreError> {
+        let filename = Self::filename(bucket.as_str(), key);
+        tracing::trace!(
+            "Fetching data from S3 for key {filename} from bucket {}",
+            self.bucket_prefix
+        );
+
+        let get_object_output = self
+            .client
+            .get_object()
+            .bucket(self.bucket_prefix.clone())
+            .key(filename)
+            .send()
+            .await?;
+        Ok(get_object_output.body.collect().await?.to_vec())
+    }
+
+    async fn put_raw(
+        &self,
+        bucket: Bucket,
+        key: &str,
+        value: Vec<u8>,
+    ) -> Result<(), ObjectStoreError> {
+        let filename = Self::filename(bucket.as_str(), key);
+        tracing::trace!(
+            "Storing data to S3 for key {filename} from bucket {}",
+            self.bucket_prefix
+        );
+
+        let length = i64::try_from(value.len()).context("Object is way too big")?;
+        self.client
+            .put_object()
+            .bucket(self.bucket_prefix.clone())
+            .key(&filename)
```

Comment:
Nit: Since the API requires `impl Into<String>` here and `filename` is not used below, it would be more efficient to pass it by value: `key(filename)`.

Reply:
Done